### PR TITLE
[IRBuilder] Fold binary intrinsics

### LINKED ISSUE #61240: [IR] IRBuilderBase::CreateBinaryIntrinsic - attempt constant folding
c411965820eb803dd7eac39f80357cad663b7ba0 has to handle constant folding of many common intrinsics because `IRBuilderBase` doesn't currently handle it:

`Intrinsic::smax`
`Intrinsic::smin`
`Intrinsic::umax`
`Intrinsic::umin`
`Intrinsic::maxnum`
`Intrinsic::minnum`

Can we add a suitable `IRBuilderFolder::FoldBinaryIntrinsics` function?

### PR DESCRIPTION
**Pull request by agentcooper**
Fixes https://github.com/llvm/llvm-project/issues/61240.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
The high level approach here should be to not hard-code constant folding for specific intrinsics, but rather make use of generic APIs like ConstantFoldCall or simplifyCall.

This is not entirely straightforward because these APIs currently require an already existing call, so some refactoring would be necessary to allow calling them with e.g. just a function, function type and arguments or something like that.

**Comment by agentcooper** [comment]
@nikic Thanks for the guidance!

I've just pushed a commit where I am able to reuse `simplifyBinaryIntrinsic` in case of the `InstSimplifyFolder`. I am still not entirely sure what should I do in the `ConstantFolder` – the `ConstantFoldCall` you've mentioned is a part of Analysis and the ConstantFolder is in the IR.

**Comment by agentcooper** [comment]
I've implemented `FoldBinaryIntrinsic` in all 3 folders: ConstantFolder, TargetFolder, InstSimplifyFolder.

In case of the TargetFolder I split the `ConstantFoldScalarCall2` into 2 functions: `ConstantFoldLibCall2` that handles `LibFunc` and `ConstantFoldIntrinsicCall2` that handles intrinsics and is reused.

I've also switched the return type for `CreateBinaryIntrinsic` from `CallInst` to `Value` to reflect the fact that it can be folded.

In case of ConstantFolder I've created a new `ConstantFoldBinaryIntrinsicInstruction` which now has the code that was previously in SLPVectorizer.

Not sure why Windows build is failing, but Linux one is passing.

@nikic Could you please take another look?

**Comment by agentcooper** [comment]
@nikic ping :-)

File: llvm/include/llvm/Analysis/ConstantFolding.h
Comment (dtcxzyw) [inline llvm/include/llvm/Analysis/ConstantFolding.h:27]:
```suggestion
using ID = unsigned;
```

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp
Comment (dtcxzyw) [inline llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:786]:
```suggestion
```

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp
Comment (dtcxzyw) [inline llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:773]: Please remove these debug printing statements.

File: llvm/lib/IR/ConstantFold.cpp
Comment (dtcxzyw) [inline llvm/lib/IR/ConstantFold.cpp:1763]: No newline at the end of the file
```suggestion
}

```

File: llvm/include/llvm/Analysis/InstSimplifyFolder.h
Comment (dtcxzyw) [inline llvm/include/llvm/Analysis/InstSimplifyFolder.h:122]: `LHS->getType()` is incorrect for `@llvm.is.fpclass`. I don't know why `abs/ctlz/cttz/is.fpclass` are treated as binary intrinsics.

File: llvm/lib/IR/IRBuilder.cpp
Comment (dtcxzyw) [inline llvm/lib/IR/IRBuilder.cpp:924]: Can we pass the argument `FMFSource` into `Folder.FoldBinaryIntrinsic`?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp
Comment (RKSimon) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:2260]: Value *

File: llvm/include/llvm/Analysis/InstSimplifyFolder.h
Comment (agentcooper) [reply llvm/include/llvm/Analysis/InstSimplifyFolder.h]: I've updated the `FoldBinaryIntrinsic` definition to include a `Type *` parameter, which is be passed from `CreateBinaryIntrinsic` using `F->getReturnType()`.

File: llvm/lib/IR/IRBuilder.cpp
Comment (agentcooper) [reply llvm/lib/IR/IRBuilder.cpp]: It is not immediately clear to me how it can be used, given that `FoldBinaryIntrinsic` returns a `Constant` and from my understanding the `FMFSource` is used to copy FMF flags when a new `CallInst` is created.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
Comment (agentcooper) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Nice! Done.

File: llvm/lib/IR/IRBuilder.cpp
Comment (dtcxzyw) [reply llvm/lib/IR/IRBuilder.cpp]: We can use FMF flags to further fold expressions in `InstSimplifyFolder`.
E.g., `minimum(X, -inf) -> -inf when FMF nnan is set`.

File: llvm/include/llvm/Analysis/InstSimplifyFolder.h
Comment (dtcxzyw) [inline llvm/include/llvm/Analysis/InstSimplifyFolder.h:121]:
```suggestion
                             Instruction *FMFSource) const override {
```
Don't use default argument in an overridden virtual function.

File: llvm/include/llvm/Analysis/TargetFolder.h
Comment (dtcxzyw) [inline llvm/include/llvm/Analysis/TargetFolder.h:195]:
```suggestion
                             Instruction *FMFSource) const override {
```

File: llvm/include/llvm/Analysis/TargetFolder.h
Comment (dtcxzyw) [inline llvm/include/llvm/Analysis/TargetFolder.h:199]: It doesn't help constant folding :(

File: llvm/include/llvm/IR/ConstantFolder.h
Comment (dtcxzyw) [inline llvm/include/llvm/IR/ConstantFolder.h:187]:
```suggestion
                             Instruction *FMFSource) const override {
```

File: llvm/include/llvm/IR/NoFolder.h
Comment (dtcxzyw) [inline llvm/include/llvm/IR/NoFolder.h:116]:
```suggestion
                             Instruction *FMFSource) const override {
```

File: llvm/include/llvm/Analysis/TargetFolder.h
Comment (dtcxzyw) [reply llvm/include/llvm/Analysis/TargetFolder.h:199]: Oh, my bad. It is used to fold constrained fp intrinsics.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thanks!
Please wait for additional approval from other reviewers.

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp
Comment (nikic) [reply llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:780]: Why? dyn_cast should always be used with auto.

**Review by nikic — CHANGES_REQUESTED** [review]
Please remove the changes to IR/ConstantFold. I'm not willing to duplicate constant folding logic in there. If we *really* wanted to we could *move* logic there, but we should not copy it.

File: llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp
Comment (nikic) [reply llvm/lib/Target/AMDGPU/AMDGPUInstCombineIntrinsic.cpp:780]: For reference, the guideline for this is at https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable.

**Comment by agentcooper** [comment]
@nikic So the suggestion is to keep the SLPVectorizer code as it is right now and maybe create a separate PR to move SLPVectorizer to TargetFolder instead? Wouldn't you consider it an improvement to extract the existing folding code from SLPVectorizer into IR/ConstantFolder?

**Comment by RKSimon** [comment]
@agentcooper reverse-ping?

**Comment by nikic** [comment]
> @nikic So the suggestion is to keep the SLPVectorizer code as it is right now and maybe create a separate PR to move SLPVectorizer to TargetFolder instead?

Yes, exactly! (Or move SLPVectorizer to TargetFolder first.)

> Wouldn't you consider it an improvement to extract the existing folding code from SLPVectorizer into IR/ConstantFolder?

No, because it sets a bad precedent. The ConstantFolding / ConstantFold split is already annoying enough as it is, without introducing actual duplication between them.

**Comment by agentcooper** [comment]
@nikic Got it. I updated the PR to exclude changes to IR/ConstantFold and SLPVectorizer (FYI @RKSimon).

**Comment by agentcooper** [comment]
@nikic ping :-)

**Review by nikic — APPROVED** [review]
LGTM

**Comment by agentcooper** [comment]
@nikic @dtcxzyw If this has enough reviews, could you please merge it for me? I don't have commit access.

**Pull request merged**
